server/rtcd.go

```diff
@@ -215,42 +215,70 @@ func (m *rtcdClientManager) getHost(ip string) *rtcdHost {
 	return m.hosts[ip]
 }
 
-// GetHostForNewCall returns the host to which a new call should be routed.
-// It performs a simple round-robin strategy based on number of calls.
-// New calls are routed to the non-flagged host with the smaller count.
+// GetHostForNewCall returns the host to which a new call should be routed to.
+// It requests system load information from all the host available (not flagged and connected)
+// and selects the one with the lower load which is assigned for the call.
 func (m *rtcdClientManager) GetHostForNewCall() (string, error) {
 	m.mut.RLock()
 	defer m.mut.RUnlock()
 
-	var h *rtcdHost
-	var minCounter uint64
-	for _, host := range m.hosts {
+	var hostsAvailable []*rtcdHost
+	for ip, host := range m.hosts {
 		host.mut.RLock()
-		// TODO: this only takes into consideration the calls started from this
-		// instance. A better strategy would be to ask rtcd for a global count (coming soon).
-		// TODO: consider also checking if the client is currently connected.
-		if !host.flagged && (h == nil || host.callsCounter < minCounter) {
-			h = host
-			minCounter = host.callsCounter
-		}
+		flagged := host.flagged
 		host.mut.RUnlock()
+
+		offline := !host.client.Connected()
+
+		// Can't assign flagged (e.g in draining process) nor offline hosts.
+		if flagged || offline {
+			m.ctx.LogDebug("skipping host from selection",
+				"host", host.ip,
+				"flagged", fmt.Sprintf("%t", flagged),
+				"offline", fmt.Sprintf("%t", offline))
+			continue
+		}
+
+		hostsAvailable = append(hostsAvailable, m.hosts[ip])
 	}
 
-	if h == nil {
+	if len(hostsAvailable) == 0 {
 		return "", fmt.Errorf("no host available")
 	}
 
-	h.mut.Lock()
-	h.callsCounter++
-	h.mut.Unlock()
+	// Now we want to select the instance with the lowest system load average.
+	var minLoad float64
+	var hostWithMinLoad *rtcdHost
+	for i, host := range hostsAvailable {
+		info, err := host.client.GetSystemInfo()
+		if err != nil {
+			m.ctx.LogError("failed to get rtcd system info", "host", host.ip, "err", err.Error())
+			continue
+		}
+
+		m.ctx.LogDebug("got system info for rtcd host", "host", host.ip, "info", fmt.Sprintf("%+v", info))
+
+		if hostWithMinLoad == nil {
+			minLoad = info.Load.Load1
+			hostWithMinLoad = hostsAvailable[i]
+		} else if info.Load.Load1 < minLoad {
+			minLoad = info.Load.Load1
+			hostWithMinLoad = hostsAvailable[i]
+		}
+	}
+
+	// Fallback to random choice if we couldn't get system info.
+	if hostWithMinLoad == nil {
+		hostWithMinLoad = hostsAvailable[rand.Intn(len(hostsAvailable))]
+	}
```

Comment:
This and the `continue` in case of error above are to make the change backward compatible in which case we'd be using a randomized approach.